// @sls-next/s3-static-assets/dist/lib/s3.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const getMimeType_1 = __importDefault(require("./getMimeType"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const aws_sdk_1 = __importDefault(require("aws-sdk"));
exports.default = ({ bucketName, credentials }) => __awaiter(void 0, void 0, void 0, function* () {
    let s3 = new aws_sdk_1.default.S3(Object.assign({}, credentials));
    try {
        const { Status } = yield s3
            .getBucketAccelerateConfiguration({
            Bucket: bucketName
        })
            .promise();
        if (Status === "Enabled") {
            s3 = new aws_sdk_1.default.S3(Object.assign(Object.assign({}, credentials), { useAccelerateEndpoint: true }));
        }
    }
    catch (err) {
        console.warn(`Checking for bucket acceleration failed, falling back to non-accelerated S3 client. Err: ${err.message}`);
    }
    return {
        uploadFile: (options) => __awaiter(void 0, void 0, void 0, function* () {
            const { filePath, cacheControl, s3Key } = options;
            const fileBody = yield fs_extra_1.default.readFile(filePath);
            return s3
                .upload({
                Bucket: bucketName,
                Key: s3Key || filePath,
                Body: fileBody,
                ContentType: getMimeType_1.default(filePath),
                CacheControl: cacheControl || undefined
            })
                .promise();
        }),
        deleteFilesByPattern: (options) => __awaiter(void 0, void 0, void 0, function* () {
            var _a;
            const { prefix, pattern, excludePattern } = options;
            const foundKeys = [];
            let continuationToken = undefined;
            while (true) {
                const data = yield s3
                    .listObjectsV2({
                    Bucket: bucketName,
                    Prefix: prefix,
                    ContinuationToken: continuationToken
                })
                    .promise();
                const contents = (_a = data.Contents) !== null && _a !== void 0 ? _a : [];
                contents.forEach(function (content) {
                    if (content.Key) {
                        const key = content.Key;
                        if (pattern.test(key) &&
                            (!excludePattern || !excludePattern.test(key))) {
                            foundKeys.push(content.Key);
                        }
                    }
                });
                if (data.IsTruncated) {
                    continuationToken = data.NextContinuationToken;
                }
                else {
                    break;
                }
            }
            const maxKeysToDelete = 1000;
            let start = 0;
            while (start < foundKeys.length) {
                const objects = [];
                for (let i = start; i < start + maxKeysToDelete && i < foundKeys.length; i++) {
                    objects.push({
                        Key: foundKeys[i]
                    });
                }
                yield s3
                    .deleteObjects({
                    Bucket: bucketName,
                    Delete: {
                        Objects: objects
                    }
                })
                    .promise();
                start += maxKeysToDelete;
            }
        }),
        getFile: (options) => __awaiter(void 0, void 0, void 0, function* () {
            var _b;
            try {
                const data = yield s3
                    .getObject({
                    Bucket: bucketName,
                    Key: options.key
                })
                    .promise();
                return (_b = data.Body) === null || _b === void 0 ? void 0 : _b.toString("utf-8");
            }
            catch (e) {
                if (e.code === "NoSuchKey") {
                    return undefined;
                }
            }
        })
    };
});
